python/ray/serve/batching.py

```diff
@@ -183,58 +191,64 @@ async def _consume_func_generator(
                         # User's code returned sentinel. No values left
                         # for caller. Terminate iteration for caller.
-                        future.set_exception(StopAsyncIteration)
+                        _set_exception_safe(future, StopAsyncIteration)
                         next_futures.append(FINISHED_TOKEN)
                     else:
                         next_future = get_or_create_event_loop().create_future()
-                        # If the client has disconnected, the future is canceled.
-                        # We should only set result when the connection is still live.
-                        if not future.cancelled():
-                            future.set_result(_GeneratorResult(result, next_future))
+                        _set_result_safe(future, _GeneratorResult(result, next_future))
                         next_futures.append(next_future)
                 futures = next_futures
 
             for future in futures:
                 if future is not FINISHED_TOKEN:
-                    future.set_exception(StopAsyncIteration)
+                    _set_exception_safe(future, StopAsyncIteration)
         except Exception as e:
             for future in futures:
                 if future is not FINISHED_TOKEN:
-                    future.set_exception(e)
+                    _set_exception_safe(future, e)
 
     async def _process_batches(self, func: Callable) -> None:
         """Loops infinitely and processes queued request batches."""
 
         while True:
-            batch: List[_SingleRequest] = await self.wait_for_batch()
-            assert len(batch) > 0
-            self_arg = batch[0].self_arg
-            args, kwargs = _batch_args_kwargs([item.flattened_args for item in batch])
-            futures = [item.future for item in batch]
-
-            # Method call.
-            if self_arg is not None:
-                func_future_or_generator = func(self_arg, *args, **kwargs)
-            # Normal function call.
-            else:
-                func_future_or_generator = func(*args, **kwargs)
-
-            if isasyncgenfunction(func):
-                func_generator = func_future_or_generator
-                await self._consume_func_generator(func_generator, futures, len(batch))
-            else:
-                try:
-                    func_future = func_future_or_generator
-                    results = await func_future
-                    self._validate_results(results, len(batch))
-                    for result, future in zip(results, futures):
-                        # If the client has disconnected, the future is canceled.
-                        # We should only set result when the connection is still live.
-                        if not future.cancelled():
-                            future.set_result(result)
-                except Exception as e:
-                    for future in futures:
-                        future.set_exception(e)
+            try:
```

Comment:
To make it clear that we are re-running forever in case of any exception, I'd suggest moving the actual logic into another method and have a top-level method that just has the `while True` and `try: except`, i.e.,
```python
def _process_batches_forever(self):
    """TODO: comment highlighting that this must catch all exceptions."""
    while True:
        try:
            self._process_batch()
        except Exception as e:
            # log exception
```

This will also help with excessive indentation.

Reply:
Good idea, I created a `_process_batches_forever` task to handle this.